jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java

```diff
@@ -197,242 +189,92 @@
+        // control characters
+        ENCODE_PATH_NEEDS_ENCODING[0x7f] = true;
+        for (int i = 0; i < 0x20; i++)
+            ENCODE_PATH_NEEDS_ENCODING[i] = true;
+    }
+
     /**
      * Encode a URI path.
      * This is the same encoding offered by URLEncoder, except that
-     * the '/' character is not encoded.
+     * the '{@code /}' character is not encoded.
      *
-     * @param path The path the encode
+     * @param path The path to encode
      * @return The encoded path
      */
     public static String encodePath(String path)
     {
         if (path == null || path.length() == 0)
             return path;
 
-        StringBuilder buf = encodePath(null, path, 0);
-        return buf == null ? path : buf.toString();
-    }
-
-    /**
-     * Encode a URI path.
-     *
-     * @param path The path the encode
-     * @param buf StringBuilder to encode path into (or null)
-     * @return The StringBuilder or null if no substitutions required.
-     */
-    public static StringBuilder encodePath(StringBuilder buf, String path)
-    {
-        return encodePath(buf, path, 0);
-    }
-
-    /**
-     * Encode a URI path.
-     *
-     * @param path The path the encode
-     * @param buf StringBuilder to encode path into (or null)
-     * @return The StringBuilder or null if no substitutions required.
-     */
-    private static StringBuilder encodePath(StringBuilder buf, String path, int offset)
-    {
-        byte[] bytes = null;
-        if (buf == null)
+        boolean needsEncoding = false;
+        boolean needsByteEncoding = false;
+        int length = path.length();
+        for (int i = 0; i < length; i++)
         {
-            loop:
-            for (int i = offset; i < path.length(); i++)
-            {
-                char c = path.charAt(i);
-                switch (c)
-                {
-                    case '%':
-                    case '?':
-                    case ';':
-                    case '#':
-                    case '"':
-                    case '\'':
-                    case '<':
-                    case '>':
-                    case ' ':
-                    case '[':
-                    case '\\':
-                    case ']':
-                    case '^':
-                    case '`':
-                    case '{':
-                    case '|':
-                    case '}':
-                        buf = new StringBuilder(path.length() * 2);
-                        break loop;
-                    default:
-                        if (c < 0x20 || c >= 0x7f)
-                        {
-                            bytes = path.getBytes(URIUtil.__CHARSET);
-                            buf = new StringBuilder(path.length() * 2);
-                            break loop;
-                        }
-                }
-            }
-            if (buf == null)
-                return null;
+            char c = path.charAt(i);
+            if (c > 0x7F) // 8-bit +
```

Comment:
0x7F should be encoded too, so `c >= 0x7F` here.

Reply:
already covered in ENCODE_PATH_NEEDS_ENCODING

Reply:
added testcase specific for this to show it's protected

Reply:
Doh, you're right.